tensorrt_llm/_torch/pyexecutor/py_executor.py

```diff
@@ -1684,14 +1684,19 @@ def _prepare_draft_tokens(self, scheduled_requests: ScheduledRequests):
             hidden_states = spec_metadata.get_hidden_states(
                 draft_batch, num_rejected_tokens)
 
-            extra_model_inputs = {'hidden_states': hidden_states}
-
             if spec_metadata.spec_dec_mode.is_eagle3():
-                # Another eagle3 hack. Eagle3 checkpoints don't have embed_tokens,
-                # so we need to provide them some other way. We can get rid of this
-                # hack if we provide our own preprocessed eagle3 checkpoints.
-                extra_model_inputs[
-                    'embed_tokens'] = self.model_engine.model.model.embed_tokens
+                hidden_states = hidden_states.to(
+                    self.draft_model_engine.model.model.dtype)
+                # Hack for eagle3. We might need to run a matmul to reduce
+                # the dimensionality of the hidden states on the first pass
+                # through the draft model. Shape dependent control flow will
+                # not work with CUDA graphs. So we just do it here.
+                expected_hidden_size = self.draft_model_engine.model.model.hidden_size
+                fc = self.draft_model_engine.model.model.fc
+                if hidden_states.shape[-1] != expected_hidden_size:
+                    hidden_states = fc(hidden_states)
```

Comment:
Can you create a function for this in the eagle model and call that function here? It's weird to have model code here.